### ISSUE TITLE
mkosi v15+: mkosi.extra/boot/ files missing in /boot, breaks incremental update_existing_rootfs()

### ISSUE THREAD

**Issue by marc-hb** [pmem/run_qemu#88]
`update_existing_rootfs()` currently relies on `/boot/System.map-N.M` being located on the main partition. When it's not, the "incremental" build fails like this:

```
not found: ./qbuild/mnt/boot/System.map-6.12.0. Try rebuilding with '-r img'
```

The `-r img` workaround is correct but obviously much slower.

Note there are multiple places where the ESP partition can be mounted: notably `/efi` or `/boot`. Fedora+mkosi seems to always use `/efi` by default?

https://wiki.archlinux.org/title/EFI_system_partition#Typical_mount_points

cc:
- #75

**Comment by marc-hb** [comment]
I can reproduce as early as mkosi v15. This was likely caused by the v15 switch to `systemd-repart`:

https://github.com/systemd/mkosi/commit/8bbbd836078a2 "Migrate disk image building to systemd-repart"

**Comment by marc-hb** [comment]
> I can reproduce as early as mkosi v15. 

Correction: with mkosi v22, `System.map-6.12.0` and friends land in the ESP partition.

With mkosi v15, they land NOWHERE!

**Comment by marc-hb** [comment]
I think we just need a `systemd-repart` configuration. It felt great to avoid one and just rely entirely on mkosi defaults but that's just too "volatile" and unpredictable for something like `update_existing_rootfs()`. Even if `update_existing_rootfs()` could get smarter and dynamically adjust its `System.map` logic now to various partition schemes, it would break again somewhere else or for some other, random mkosi version. So let's just bite the `systemd-repart` configuration bullet. I took a look and it does not look like rocket science. Also, it's still possible to leave a lot of things as default in such a configuration.

**Comment by marc-hb** [comment]
> I think we just need a systemd-repart configuration.

... or maybe not. Maybe that's not required after all...

One burning question is: what is the `-F System.map` trying to achieve? It came with the addition of the `depmod` invocation in commit 2ed0ed3af4fa2f3ec. `man depmod` says:

```
       -F, --filesyms System.map
           Supplied with the System.map produced when the kernel was built, this allows the -e
           option to report unresolved symbols. This option is mutually incompatible with -E.

```

But `-e` is currently not used! So, `-F` does nothing at all ?

Also: when invoked by `update_existing_rootfs()`, `setup_depmod()` seem to look at the OLD `System.map` file? This re-enforces the suspicion that it does nothing :-D

Could this -F be another instance of trying to port to mkosi v15+ another (#76) `update_existing_rootfs()` feature that never actually worked with v14- in the first place? If yes then let's just (temporarily) delete it to unblock the migration to v15+

Generally speaking, porting  to mkosi v15+ is really hard without a clear picture of what: 1) code was supposed to do with mkosi v14- in the first place 2) what it was _actually_ achieving with v14-.

Other complications:
The kernel and the initrd live in potentially 3 different places
  * `mkosi.extra` 
  * ESP partition
  * Root partition

--no-direct-kernel affects which are used.

On top of that, modules are also in /lib/modules.

**Comment by marc-hb** [comment]
Simply dropping the `-F System.map` argument is enough to build and boot with mkosi v15 https://github.com/pmem/run_qemu/actions/runs/12402741008/job/34624880942?pr=90

@stellarhopper , @weiny2 could you test that `-F System.map`  drop more extensively? I mean with some actual kernel and module changes...

**Comment by marc-hb** [comment]
I did a lot more testing and dropping "-F System.map" is not good enough. It's just shooting the messenger. It's a "also guilty" messenger but still just a messenger. Dropping "-F System.map" fixes the build but hides a bigger `/boot` problem:

=>  Still no kernel _modules_ in /boot!

Here's the situation with mkosi v15+ if we drop "-F System.map" 

1. run_qemu.sh from scratch; invokes `mkosi`:
  `/boot/` is totally empty
2. run_qemu.sh not from scratch, no `mkosi` but `update_init_rootfs()` intead:
  `/boot/` has the latest vmlinuz but NO KERNEL MODULES!!

The above tested with both v15 and v23.

I think it's better to fail with this "system.map" error message because it can lead people to this bug and issue until the real `/boot/ problem is actually fixed rather than _silently_ give them an empty and then mostly empty `/boot/` while pretending everything looks fine.

**Comment by stellarhopper** [comment]
Hm, didn't mean to close this - I guess it auto-closed because of the mention in #98

**Comment by marc-hb** [comment]
> I guess it auto-closed because of the mention in https://github.com/pmem/run_qemu/pull/98

Most likely yes, please upvote https://github.com/orgs/community/discussions/17308 (and duplicates...)

**Comment by marc-hb** [comment]
> Dropping "-F System.map" fixes the build but hides a bigger missing `/boot` problem.

So the key question is: does anyone or anything uses `/boot`?

`/boot` inside the image is used by neither `--direct-kernel` nor by `--no-direct-kernel` right now. The former uses the kernel and initrd _outside_ the image. The latter uses the `/efi` partition.

Maybe `/boot/` was used in older, GRUB times but not anymore now? @stellarhopper, @weiny2 , any memories?

If `/boot` is not used or not used anymore, then we can drop `/boot` entirely, point `-F System.map` somewhere else and the problem should be solved!

**Comment by stellarhopper** [comment]
@marc-hb yeah I'm pretty sure this is true - /boot is just a holdover from grub days, and likely can be removed now.

**Comment by marc-hb** [comment]
> update_existing_rootfs() currently relies on /boot/System.map-N.M being located on the main partition. When it's not, the "incremental" build fails like this:...
> I can reproduce as early as mkosi v15. 

I don't understand how this stopped being an issue. Was it Fedora 40 specific? I'm not using Fedora much these days.
